lib/semantics/check-do-stmt.cc

```diff
@@ -344,41 +372,101 @@ static CS GatherReferencesFromExpression(const parser::Expr &expression) {
   }
 }
 
-// Find a canonical DO CONCURRENT and enforce semantics checks on its body
-class DoConcurrentContext {
+// Find a DO statement and enforce semantics checks on its body
+class DoStmtContext {
 public:
-  DoConcurrentContext(SemanticsContext &context)
-    : messages_{context.messages()} {}
-
-  bool operator==(const DoConcurrentContext &x) const { return this == &x; }
+  DoStmtContext(SemanticsContext &context) : context_{context} {}
 
   void Check(const parser::DoConstruct &doConstruct) {
+    if (doConstruct.IsDoConcurrent()) {
+      CheckDoConcurrent(doConstruct);
+      return;
+    }
+    if (doConstruct.IsDoNormal()) {
+      CheckDoNormal(doConstruct);
+      return;
+    }
+    // TODO: handle the other cases
+  }
+
+private:
+  using Bounds = parser::LoopControl::Bounds;
+
+  const Bounds &GetBounds(const parser::DoConstruct &doConstruct) {
+    auto &loopControl{doConstruct.GetLoopControl().value()};
+    return std::get<Bounds>(loopControl.u);
+  }
+
+  void CheckDoControl(parser::CharBlock sourceLocation, bool isReal) {
+    bool warn = context_.warnOnNonstandardUsage() ||
+        context_.ShouldWarn(parser::LanguageFeature::RealDoControls);
+    bool enabled = context_.IsEnabled(parser::LanguageFeature::RealDoControls);
+    if (isReal && enabled && !(warn)) {
+      // No messages for the default case
+    } else if (isReal && enabled && warn && !context_.warningsAreErrors()) {
+      context_.Say(sourceLocation, "DO controls should be INTEGER"_en_US);
+    } else {
+      context_.Say(sourceLocation, "DO controls should be INTEGER"_err_en_US);
+    }
+  }
+
+  void CheckDoVariable(const parser::ScalarName &scalarName) {
+    const DeclTypeSpec *symType{scalarName.thing.symbol->GetType()};
```

Comment:
Also, regarding `scalarName.thing.symbol`, maybe you should check that it is not null, though I am not completely sure it can.
@klausler and @tskeith , should `GetUltimate()` be used here also in case the do-variable is coming from use association (I am not quite sure it is legal, but cannot find why not either)?

Reply:
Those are both good points. It may be that`symbol` is null if the do variable is not declared and IMPLICIT NONE is in effect. And I also cannot find any indication that the do variable may not be use-associated.

Reply:
Thanks to both of you for pointing this out.  @tskeith, I wrote a test, and you're correct that a
variable that's not declared causes a seg fault in my current implementation.

I'm not sure if I understand the references to use association and their implications for using
`GetUltimate()`.  I don't see anything in the standard that prohibits using a variable from a
`MODULE` as a DO variable.  I wrote the following test program, and everything seems to work
as expected:
`MODULE share
  INTEGER :: intvar
  REAL :: realvar
END MODULE share

PROGRAM use_assoc
  USE share
  DO intvar = 1, 10
    PRINT *, "hello"
  END DO
  DO realvar = 1, 10
    PRINT *, "hello"
  END DO
END PROGRAM use_assoc
`

Reply:
As long as it works correctly it should be fine. Some tests would be a good idea.

It works in this case because `Symbol::GetType()` follows use associations to get the type from the ultimate symbol. To query some properties of use-associated symbols you have to call `GetUltimate()` on them and look at that symbol.

Reply:
Thanks, @tskeith .  I'll add a test.